### PR TITLE
Move death formspec and respawn to server builtin

### PR DESCRIPTION
**Pull request by pauloue**
Currently we have:
Player dies --> TOCLIENT_DEATHSCREEN --> client shows death formspec --> when respawn button clicked send a TOSERVER_RESPAWN --> server respawns player
I removed those two packets to make it:
Player dies --> server sends death formspec --> when respawn button clicked respawn the player

Since the death formspec is defined server-side, this opens up the possibility of the server customizing it (see lua_api.txt).
The default death formspec and respawn behavior are unchanged.

Some of the deleted C++ code referenced camera stuff to be used to point to the death cause, but this is not yet implemented. If/when it is done, it could probably be done in builtin.

TODO: `find_spawn_pos` should throw an error if called before the map is loaded (like `get_spawn_level`).

### CODE REVIEW

File: builtin/game/respawn.lua [hunk 1]
```diff
@@ -0,0 +1,39 @@
+-- Minetest: builtin/game/respawn.lua
+
+function core.show_death_formspec(player)
+	core.show_formspec(player:get_player_name(), "builtin:death", [[
```
Comment (SmallJoker) [inline builtin/game/respawn.lua]: This allows any player to respawn instantly by sending this formspec to the server. 
The code in `register_on_player_receive_fields` may only be executed when the player died but did not respawn yet.

Comment (SmallJoker) [reply builtin/game/respawn.lua]: Also: This is just a regular formspec. Delayed `show_formspec` calls may override this one; so the player cannot respawn.

What if the player joined with 0 HP? They're dead. Show the respawn dialogue to them.

Comment (pauloue) [reply builtin/game/respawn.lua]: Nice, I forgot to test an important thing. Will investigate.

**Comment by paramat** [comment]
Does the current `minetest.send_respawn()` allow client-provided CSM to respawn the player? I'm wondering why that is allowed, surely the server or server mods should control whehter a player can respawn?

**Comment by pauloue** [comment]
It does allow the client to respawn the player, but it only works if the player is dead. So it allows skipping the death screen with this bit of CSM: `minetest.register_on_death(minetest.send_respawn)`.

Note that this PR is somewhat experimental and may have some issues.

File: builtin/game/respawn.lua (hunk 1 above)
Comment (pauloue) [reply builtin/game/respawn.lua]: Fixed the first comment.
> What if the player joined with 0 HP?

The server sends the HP when a player joins, so on_dieplayer will be triggered (tested).

**Comment by nerzhul** [comment]
Sorry but i'm not in favor of this in term of roadmap, for the 5.0.0. We are so close to freeze i prefer to delay. there is no need to have this as a blocker, this is not an awaited feature.

**Comment by nerzhul** [comment]
Also this is a very sensible feature in terms of workflow and can permit easily exploits if problems in the workflow

**Comment by pauloue** [comment]
Yes, and there's a pretty big issue with the formspec. I'd like to know what the possible exploits are though.

**Comment by paramat** [comment]
@nerzhul 
> Also this is a very sensible feature in terms of workflow

Do you mean 'sensitive'? I've seen you use 'sensible' elsewhere where 'sensitive' makes more sense.
>  We are so close to freeze i prefer to delay. there is no need to have this as a blocker, this is not an awaited feature which is required to 5.0.0

It was marked as blocker because it changes network compatibility, so would have to be for 5.0.0.
> this is not an awaited feature which is required to 5.0.0

Says who? At any point we can decide that something would be good, or is essential for, 5.0.0.

However, thinking on this i do agree with you and SmallJoker and don't support this now, because i don't see much need for modifying the deathscreen either clientside or serverside, and this could introduce subtle bugs.
Because this changes network compatibility we would then be 'locked in' to losing those 2 packets forthe whole of 5.0.0, too risky.
:-1:

File: builtin/game/respawn.lua (hunk 1 above)
Comment (pauloue) [reply builtin/game/respawn.lua]: > This is just a regular formspec.

I thought all formspecs were the same. What's a non-regular formspec?

> Delayed `show_formspec` calls may override this one; so the player cannot respawn.

I did some testing and it seems the current death formspec (both the C++ and CSM versions) has the same issue.

**Comment by nerzhul** [comment]
yeah my english is not perfect. I would mean this move is very interesting but can trigger many security and gameplay issues with servers, i prefer to use the 5.1.0 dev cycle to push it and take time to test it properly as we are very close to the 5.0.0 freeze.

**Comment by jastevenson303** [comment]
I think that this is related to the issue with a player moving around when they're dead.  This formspec seems to prohibit it, and when it doesn't show (as sometimes happens) then the player can still move around while hp == 0.

**Comment by SmallJoker** [comment]
Hopelessly drown in rebase hell.
Would it be possible to split the server->builtin code movement and the `DiePlayer` changes? I think visuals and the actual death code should be possible to separate for easier testing and reviewing.

**Comment by sfan5** [comment]
I think this PR could be much simplified by adding a formspec to `TOCLIENT_DEATHSCREEN` and keeping `TOSERVER_RESPAWN` for use with a special formspec button.
This would also be easier to implement in a backwards compatible way.

**Comment by pauloue** [comment]
I did manage to rebase it.
It works with a new server and old client. The client doesn't show its death screen because the server doesn't send `TOCLIENT_DEATHSCREEN`. The server death screen is shown because it just uses `minetest.show_formspec`.
No death formspec is show with an old server and new client, but the only way to keep backwards compatibility in that case is to keep the client death formspec.
So I'm not sure it's worth doing this without breaking backwards compatibility.

**Comment by Wuzzy2** [comment]
Cool stuff.

I have one complaint:

```
core.show_formspec(player:get_player_name(), "builtin:death",
```

This seems to be the first `builtin` formspec ever so we might as well settle on a naming convention first. I think `__builtin:` should be the prefix instead of `builtin:`, in order to parallel the builtin entities.

**Comment by Wuzzy2** [comment]
Tested, works nicely. I really like the new API functions as well. I just now realized how much the client-side implementation sucks, moving it to server side opens so many new possibilities. It is long overdue to move this stuff away from the client side, so good job. :+1:

Unfortunately, this PR seems to have outdated *again* with 2 file conflicts (our review process sucks); hopefully it's not too bad.

I noticed the death formspec is larger than before, but it's not important. It uses no `formspec_version` yet, but I suggets to ignore that fact for this PR, and deal with that later; there are other "core formspecs" in Minetest with missing `formspec_version`.

I suggest to change "You died" to "You died." cuz it's a full English sentence; the missing period has always bothered me. xD

**Comment by AFCMS** [comment]
I support this idea. Customizing death formspec is really needed.

**Comment by LoneWolfHT** [comment]
Is a rebase the only thing blocking this?

**Pull request closed without merging**